### PR TITLE
feat: add SMTP auth & TLS support

### LINKED ISSUE #1: Implement auth & TLS for SMTP notification dispatcher
Acceptance criteria:

- [x] `PLAIN` auth support
- [x] `LOGIN` auth support
- [x] TLS support (including `STARTTLS`)
- [x] Can send a notification to **GSuite** (may require additional SASL auth mechanism)
- [x] Can send a notification to **Office 365** (may require additional SASL auth mechanism)

### PR DESCRIPTION
**Pull request by dannykopping**
Closes https://github.com/coder/team-coconut/issues/1
Partially addresses https://github.com/coder/team-coconut/issues/5

Also changed the email template to use a hosted logo since [SVG support in email clients is still pitiful](https://css-tricks.com/a-guide-on-svg-support-in-email/).

### CODE REVIEW

File: codersdk/deployment.go [hunk 1]
```diff
@@ -2121,7 +2158,7 @@ Write out the current server config as YAML to stdout.`,
 			Value:       &c.Notifications.DispatchTimeout,
 			Default:     time.Minute.String(),
 			Group:       &deploymentGroupNotifications,
-			YAML:        "dispatch-timeout",
+			YAML:        "dispatchTimeout",
```
Comment (dannykopping) [inline codersdk/deployment.go:2162]: Drive-by: fixing previous mistakes of mine.

File: .github/workflows/typos.toml [hunk 2]
```diff
@@ -20,6 +20,8 @@ hel = "hel"
 pn = "pn"
 # typos doesn't like the EDE in TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA
 EDE = "EDE"
+# HELO is an SMTP command
+HELO = "HELO"
```
Comment (mtojek) [inline .github/workflows/typos.toml:24]: BTW is it HELO not EHLO?

File: cli/testdata/server-config.yaml.golden [hunk 3]
```diff
@@ -493,13 +493,15 @@ userQuietHoursSchedule:
...
 allowWorkspaceRenames: false
+# Configure how notifications are processed and delivered.
 notifications:
   # Which delivery method to use (available options: 'smtp', 'webhook').
   # (default: smtp, type: string)
   method: smtp
   # How long to wait while a notification is being sent before giving up.
   # (default: 1m0s, type: duration)
-  dispatch-timeout: 1m0s
+  dispatchTimeout: 1m0s
```
Comment (mtojek) [inline cli/testdata/server-config.yaml.golden:503]: Looks like we're mixing _camelCase_ and _snake_case_ in the config. Should we stick with one format?

File: coderd/notifications/dispatch/smtp.go [hunk 4]
```diff
@@ -133,47 +131,42 @@ func (s *SMTPHandler) dispatch(subject, htmlBody, plainBody, to string) Delivery
...
+			return false, xerrors.New("no authentication mechanisms supported by server")
+		}
 
 		// Sender identification.
 		from, err := s.validateFromAddr(s.cfg.From.String())
 		if err != nil {
 			return false, xerrors.Errorf("'from' validation: %w", err)
 		}
-		err = c.Mail(from)
+		err = c.Mail(from, &smtp.MailOptions{})
```
Comment (mtojek) [inline coderd/notifications/dispatch/smtp.go:157]: Do you think we can build a fluent interface? Something like:

`c.Mail().Rcpts(...).Data(...).Send()`

File: coderd/notifications/dispatch/smtp.go [hunk 5]
```diff
@@ -93,37 +104,24 @@ func (s *SMTPHandler) dispatch(subject, htmlBody, plainBody, to string) Delivery
...
-		conn, err = d.DialContext(ctx, "tcp", fmt.Sprintf("%s:%s", smarthost, smarthostPort))
-		if err != nil {
-			return true, xerrors.Errorf("establish connection to server: %w", err)
+		if _, ok := ctx.Deadline(); !ok {
+			return false, xerrors.Errorf("context has no deadline")
 		}
 
-		// Create an SMTP client.
-		c, err = smtp.NewClient(conn, smarthost)
+		// TODO: reuse client across dispatches (if possible).
```
Comment (mtojek) [inline coderd/notifications/dispatch/smtp.go:120]: I'm afraid we have to implement it now (or in follow-up as this PR is large). SMTP servers may have enabled restrictions limiting login attempts.

**Review by mtojek — COMMENTED** [review]
Have you done a trial with Google account?

File: coderd/notifications/dispatch/smtp.go (hunk 5 above)
Comment (dannykopping) [reply coderd/notifications/dispatch/smtp.go:120]: I doubt it; Alertmanager does it this way and I'm not aware of any issues arising from that. I don't think we'll need this for the initial version.

File: coderd/notifications/dispatch/smtp.go (hunk 4 above)
Comment (dannykopping) [reply coderd/notifications/dispatch/smtp.go:157]: We could consider it, but I don't personally see the value. I'm very willing to be convinced though.

File: coderd/notifications/dispatch/smtp.go (hunk 5 above)
Comment (mtojek) [reply coderd/notifications/dispatch/smtp.go:120]: I see, so it will just operate slower. Let's leave it then.

File: coderd/notifications/dispatch/smtp.go (hunk 4 above)
Comment (mtojek) [reply coderd/notifications/dispatch/smtp.go:157]: Readability, but I won't push for it. This syntax is also fine.

File: coderd/notifications/dispatch/smtp.go [hunk 6]
```diff
@@ -260,10 +253,214 @@ func (s *SMTPHandler) dispatch(subject, htmlBody, plainBody, to string) Delivery
...
+
+			return sasl.NewPlainClient(s.cfg.Auth.Identity.String(), username, password), nil
+		case sasl.Login:
+			if slices.Contains(list, sasl.Plain) {
+				// Prefer PLAIN over LOGIN.
+				continue
+			}
+
+			// Warn that LOGIN is obsolete, but don't do it every time we dispatch a notification.
+			loginWarnOnce.Do(func() {
```
Comment (mtojek) [inline coderd/notifications/dispatch/smtp.go:442]: nit: Should we move it to initialization of the dispatcher?

File: coderd/notifications/dispatch/smtp.go (hunk 5 above)
Comment (dannykopping) [reply coderd/notifications/dispatch/smtp.go:120]: Added https://github.com/coder/team-coconut/issues/15

File: .github/workflows/typos.toml (hunk 2 above)
Comment (dannykopping) [reply .github/workflows/typos.toml:24]: Both. If EHLO is not supported the library falls back to HELO.

File: cli/testdata/server-config.yaml.golden (hunk 3 above)
Comment (dannykopping) [reply cli/testdata/server-config.yaml.golden:503]: Whoops, you're right. Thanks 👍 
We have a general problem with this it seems, but I'll at least try to make all of my additions internally consistent.

**Review by mtojek — APPROVED** [review]
👍

File: coderd/notifications/dispatch/smtp.go (hunk 6 above)
Comment (dannykopping) [reply coderd/notifications/dispatch/smtp.go]: Good idea 👍

**Comment by dannykopping** [comment]
@mtojek heads up, added this after your review: https://github.com/coder/coder/pull/13902/commits/4f7a91024720d943be19c85dd7a30ed4494f77fe

Changed the email template to use a hosted logo since [SVG support in email clients is still pitiful](https://css-tricks.com/a-guide-on-svg-support-in-email/).

**Comment by mtojek** [comment]
> Changed the email template to use a hosted logo since [SVG support in email clients is still pitiful](https://css-tricks.com/a-guide-on-svg-support-in-email/).

what about embedded PNG/base64 image?

**Comment by dannykopping** [comment]
> > Changed the email template to use a hosted logo since [SVG support in email clients is still pitiful](https://css-tricks.com/a-guide-on-svg-support-in-email/).
> 
> what about embedded PNG/base64 image?

Tried that too, it shows like this:

image

**Comment by mtojek** [comment]
Crap, that's bad. I'm wondering if we shouldn't host it on our site rather than on GitHub. I don't know if these raw links are stable API.

**Comment by dannykopping** [comment]
I've created https://github.com/coder/coder.com/pull/518 and referenced the new URL in https://github.com/coder/coder/pull/13902/commits/5a8331f53016260787b29cd17e0c3d568783e7ff
Once that's merged I'll merge this one
